presto-main/src/main/java/com/facebook/presto/cost/HistoryBasedPlanStatisticsCalculator.java

```diff
@@ -38,42 +48,92 @@
+            }
         }
-        List<PlanNode> children = planNode.getSources().stream().map(node -> removeGroupReferences(node, lookup)).collect(toImmutableList());
-        return planNode.replaceChildren(children);
+        // Cache statistics for future calls
+        // TODO: Right now, we assume that underlying implementation will cache statistics
+        // when called. We should make this explicit by adding a cache here or through API.
+        externalPlanStatisticsProvider.get().getStats(hashes.build());
+    }
+
+    @VisibleForTesting
+    public Optional<CanonicalPlanHashes> getCanonicalPlanHashes(QueryId queryId)
+    {
+        return Optional.ofNullable(planNodeHashes.get(queryId));
     }
 
-    private PlanStatistics getStatistics(PlanNode planNode, Session session, TypeProvider types, Lookup lookup)
+    private PlanStatistics getStatistics(PlanNode planNode, Session session, Lookup lookup)
     {
-        planNode = removeGroupReferences(planNode, lookup);
-        ExternalPlanStatisticsProvider externalStatisticsProvider = externalPlanStatisticsProvider.get();
+        final PlanNode plan = resolveGroupReferences(planNode, lookup);
         if (!useExternalPlanStatisticsEnabled(session)) {
             return PlanStatistics.empty();
         }
-        Map<PlanNodeWithHash, HistoricalPlanStatistics> statistics =
-                externalStatisticsProvider.getStats(ImmutableList.of(new PlanNodeWithHash(planNode, Optional.empty())));
 
-        if (statistics.size() == 1) {
-            return statistics.values().iterator().next().getLastRunStatistics();
+        Map<PlanCanonicalizationStrategy, String> allHashes = getCanonicalPlanHashes(session.getQueryId())
+                .map(hashes -> hashes.getCanonicalPlanHashes(planNode.getId()))
+                .orElseGet(ImmutableMap::of);
+
+        List<PlanNodeWithHash> planNodeWithHashes = allHashes.values().stream()
+                .map(hash -> new PlanNodeWithHash(plan, Optional.of(hash)))
+                .collect(toImmutableList());
+
+        // If no hashes are found, try to fetch statistics without hash.
+        if (planNodeWithHashes.isEmpty()) {
+            planNodeWithHashes = ImmutableList.of(new PlanNodeWithHash(plan, Optional.empty()));
+        }
+
+        Map<PlanNodeWithHash, HistoricalPlanStatistics> statistics = externalPlanStatisticsProvider.get().getStats(planNodeWithHashes);
+
+        // Return statistics corresponding to first strategy that we find, in order specified by `historyBasedPlanCanonicalizationStrategyList`
+        for (PlanCanonicalizationStrategy strategy : historyBasedPlanCanonicalizationStrategyList()) {
+            for (Map.Entry<PlanNodeWithHash, HistoricalPlanStatistics> entry : statistics.entrySet()) {
+                if (allHashes.containsKey(strategy) && Optional.of(allHashes.get(strategy)).equals(entry.getKey().getHash())) {
+                    return entry.getValue().getLastRunStatistics();
```

Comment:
I would add a todo here to update the algorithm for what statistics to return (if we have data from more runs, then be smarter than just stats from the last run)